File: 윤이진/보석쇼핑.py
def solution(gems):
    dic=dict()
    s=set(gems)
    for i in s:
        dic[i]=0
    kind=len(s)
    print(kind)
    
    gems.insert(0,'')
    l,r,cnt=1,1,1
    dic[gems[l]]=1
    answer=[1,len(gems)-1]
    
    while l<len(gems) and r<len(gems):
        if cnt<kind:
            r+=1
            if r>=len(gems): break
            
            if dic[gems[r]]==0:
                cnt+=1
            
            dic[gems[r]]+=1
        else:
            if r-l<answer[1]-answer[0]:
                answer=[l,r]
                
            if dic[gems[l]]==1:
                cnt-=1
                
            dic[gems[l]]-=1
            l+=1
            
    
    return answer

File: 안영진/보석쇼핑.java
import java.util.*;

class Solution {
    public int[] solution(String[] gems) {
        int[] answer = {};
        // 보석의 종류를 알기위해 해시맵에 넣어서 중복 제거
        HashSet <String> hs = new HashSet<>();
        int gems_length = gems.length;
        for (int i =0; i< gems_length; i++){
            hs.add(gems[i]);
        }
        
        int gems_size = hs.size();
        
        System.out.print(gems_size);
        //그 다음에 뭔가 탐색을 해야할거같은데 떠오르지가않는다ㅏㅏㅏ
        
        
        return answer;
    }
}